<a href="https://colab.research.google.com/github/grandfatherr/Electric-car-Usage/blob/main/Edwin_Mutuma_Wk_4_IP.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#1.0 Importing libraries 
These libraries enable us to access some functions that we otherwise couldn't access.
For this, we will import pandas and numpy

In [1]:
import pandas as pd
import numpy as np


#1.1 Previewing the dataset

In [2]:
elec=pd.read_csv('http://bit.ly/autolib_dataset')
elec.head()

Address  Cars  Bluecar counter  ...  day  hour  minute
0                 2 Avenue de Suffren     0                0  ...    8    11      43
1           145 Rue Raymond Losserand     6                6  ...    6     7      24
2    2 Avenue John Fitzgerald Kennedy     3                3  ...    3    20      14
3              51 Rue Eugène Oudiné     3                3  ...    4     4      37
4  6 avenue de la Porte de Champerret     3                3  ...    8    17      23

[5 rows x 25 columns]

In [3]:
elec.sample(10)

Address  Cars  Bluecar counter  ...  day  hour  minute
4723                   6 Rue des Bas     3                3  ...    8    11      55
603               2 rue Paul Dautier     7                7  ...    7    13      45
2243     4 rue Alexis de Tocqueville     1                1  ...    2    11       9
1943                   52 Rue Carnot     3                3  ...    3     8      18
3905           1 rue Olympe de Gouge     4                4  ...    9    11      33
1851            178 Avenue Galliéni     3                3  ...    5     4      16
3827            19 rue Léon Renault     6                6  ...    9     7      31
1193             5 Rue Henry Brisson     1                1  ...    2     2      22
2692        22 place Charles Fillion     4                4  ...    5    23      28
3693  24 Boulevard de la République     2                2  ...    9     0      16

[10 rows x 25 columns]

In [4]:
elec.tail()

Address  Cars  Bluecar counter  ...  day  hour  minute
4995          8 avenue Ménelotte     2                2  ...    6    11      26
4996            37 rue de Dantzig     4                4  ...    4    16      56
4997               142 rue du Bac     1                1  ...    1     7       1
4998  2 avenue du Val de Fontenay     2                2  ...    4    17      27
4999          5 Rue Henry Brisson     4                4  ...    8    10       2

[5 rows x 25 columns]

#1.2 Accessing information on our dataset
Let's look at the information and the description of our dataset

In [5]:
elec.info

<bound method DataFrame.info of                                  Address  Cars  ...  hour  minute
0                    2 Avenue de Suffren     0  ...    11      43
1              145 Rue Raymond Losserand     6  ...     7      24
2       2 Avenue John Fitzgerald Kennedy     3  ...    20      14
3                 51 Rue Eugène Oudiné     3  ...     4      37
4     6 avenue de la Porte de Champerret     3  ...    17      23
...                                  ...   ...  ...   ...     ...
4995                 8 avenue Ménelotte     2  ...    11      26
4996                   37 rue de Dantzig     4  ...    16      56
4997                      142 rue du Bac     1  ...     7       1
4998         2 avenue du Val de Fontenay     2  ...    17      27
4999                 5 Rue Henry Brisson     4  ...    10       2

[5000 rows x 25 columns]>

In [6]:
elec.describe

<bound method NDFrame.describe of                                  Address  Cars  ...  hour  minute
0                    2 Avenue de Suffren     0  ...    11      43
1              145 Rue Raymond Losserand     6  ...     7      24
2       2 Avenue John Fitzgerald Kennedy     3  ...    20      14
3                 51 Rue Eugène Oudiné     3  ...     4      37
4     6 avenue de la Porte de Champerret     3  ...    17      23
...                                  ...   ...  ...   ...     ...
4995                 8 avenue Ménelotte     2  ...    11      26
4996                   37 rue de Dantzig     4  ...    16      56
4997                      142 rue du Bac     1  ...     7       1
4998         2 avenue du Val de Fontenay     2  ...    17      27
4999                 5 Rue Henry Brisson     4  ...    10       2

[5000 rows x 25 columns]>

#1.3 Data cleaning.
The data that we are presented with most of the times will not be clean data. Now we know the saying garbage in, garbage out. It'd not be possible to use unclean data and expect great and useable insight.

Data cleaning is the process of detecting and correcting corrupt and inaccurate values and records from a dataset as we are going to see below. We will be identifying incomplete, incorrect, inaccurate and or irrelevant values in our data and then replace, modify and or deleting the dirty values in the data.

##1.31 Validity


In [7]:
# Looking at our data there are columns that we are not going to use that will 
# need to be dropped.
#
ev = elec.drop(['Cars', 'ID', 'Geo point', 'year', 'month'], axis = 1, inplace=False)

In [8]:
# Let's look at our dataframe after dropping the columns
#
ev.head()

Address  Bluecar counter  ...  hour  minute
0                 2 Avenue de Suffren                0  ...    11      43
1           145 Rue Raymond Losserand                6  ...     7      24
2    2 Avenue John Fitzgerald Kennedy                3  ...    20      14
3              51 Rue Eugène Oudiné                3  ...     4      37
4  6 avenue de la Porte de Champerret                3  ...    17      23

[5 rows x 20 columns]

##1.32 Accuracy

##1.33 Completeness

In [9]:
# Let's now look at whether all the data values in our dataframe is available 
# to work with.
#
ev.isnull().sum()

Address                   0
Bluecar counter           0
Utilib counter            0
Utilib 1.4 counter        0
Charge Slots              0
Charging Status           0
City                      0
Displayed comment      4889
Kind                      0
Postal code               0
Public name               0
Rental status             0
Scheduled at           4953
Slots                     0
Station type              0
Status                    0
Subscription status       0
day                       0
hour                      0
minute                    0
dtype: int64

In [10]:
# Since we have two columns that have more than two thirds wof missing values,
# we are going to drop them so that we have data that we can use.
#
ev = ev.drop(['Scheduled at', 'Displayed comment'], axis = 1, inplace=False)

In [11]:
# Let us look at the table
ev.head(1)

Address  Bluecar counter  Utilib counter  ...  day  hour minute
0  2 Avenue de Suffren                0               0  ...    8    11     43

[1 rows x 18 columns]

In [12]:
# Let's check to see if our table has complete values
ev.isnull().sum()

Address                0
Bluecar counter        0
Utilib counter         0
Utilib 1.4 counter     0
Charge Slots           0
Charging Status        0
City                   0
Kind                   0
Postal code            0
Public name            0
Rental status          0
Slots                  0
Station type           0
Status                 0
Subscription status    0
day                    0
hour                   0
minute                 0
dtype: int64

##1.34 Consistency

In [13]:
# We check for consistency in our dataset by looking for any duplicates in the 
# data
#If there are any duplicates, we will drop them 
#
ev.duplicated().sum()

0

##1.35 Uniformity 

In [14]:
# Lets make the column names uniform so that we may be able to work on the same seamlessly
#
ev=ev.rename(columns=str.lower)
ev.columns = ev.columns.str.strip()


In [15]:
# Let us now preview our dataset after making these changes
#
ev.head(2)

address  bluecar counter  ...  hour  minute
0        2 Avenue de Suffren                0  ...    11      43
1  145 Rue Raymond Losserand                6  ...     7      24

[2 rows x 18 columns]

#1.4 Analysis

##1.4.1 The most popular hour of the day for picking up a shared electric car (Bluecar) in the city of Paris over the month of April 2018.

In [16]:
# Most popular hour of the day for picking up blue car
#
ev.loc[(ev['bluecar counter'] == 0) & (ev['city'] == 'Paris')]['hour'].value_counts().sort_values(ascending = False).head(2)

2     54
21    52
Name: hour, dtype: int64

##1.4.2 The most popular hour for returning cars

In [17]:
# The most popular hour for returning cars
ev[ev['slots']==0]['hour'].value_counts().head(2)

21    89
9     88
Name: hour, dtype: int64

##1.4.3 The most popular station overall

In [18]:
# most popular station overall
#
ev.groupby('station type')['public name'].value_counts().sort_values(ascending=False).head(1)

station type  public name               
station       Paris/Porte de Montrouge/8    13
Name: public name, dtype: int64

##1.4.4 The most popular postal code for picking up Blue cars? 

In [21]:
# What postal code is the most popular for picking up Blue cars? Does the most popular station belong to that postal code?
ev.loc[(ev['bluecar counter']==0) & (ev['hour']==21)]['postal code'].value_counts().head(1)

75008    7
Name: postal code, dtype: int64

#1.5 Recommendations 

We should increase the availability of blue car for picking up at the Postal Code 75008 since it was the most popular for picking up Blue car. It is also advisable to increase services that are offered at the Paris/Porte de Montrouge/8 Station since it was the most popular station.
We should also look at the prospectus of increasing the blue cars during the second hour and the 21st hours as these were found to be the most popular hours for picking up blue cars.